crates/turborepo-lib/src/daemon/connector.rs

```diff
@@ -0,0 +1,278 @@
+use std::{
+    path::{Path, PathBuf},
+    process::Stdio,
+    sync::Arc,
+    time::Duration,
+};
+
+use command_group::AsyncCommandGroup;
+use log::{debug, error};
+use notify::{Config, Event, EventKind, RecommendedWatcher, Watcher};
+use sysinfo::{ProcessExt, ProcessRefreshKind, RefreshKind, SystemExt};
+use tokio::{net::UnixStream, sync::mpsc, time::timeout};
+use tonic::transport::Endpoint;
+
+use super::{client::proto::turbod_client::TurbodClient, DaemonClient, DaemonError};
+
+#[derive(Debug)]
+pub struct DaemonConnector {
+    pub dont_start: bool,
+    pub dont_kill: bool,
+    pub pid_file: PathBuf,
+    pub sock_file: PathBuf,
+}
+
+impl DaemonConnector {
+    const CONNECT_RETRY_MAX: usize = 3;
+    const SHUTDOWN_TIMEOUT: Duration = Duration::from_secs(1);
+    const SOCKET_TIMEOUT: Duration = Duration::from_secs(1);
+
+    /// Attempt, with retries, to:
+    /// 1. find (or start) the daemon process
+    /// 2. locate its unix socket
+    /// 3. connect to the socket
+    /// 4. send the 'hello' message, negotiating versions
+    ///
+    /// A new server will be spawned (and the old one killed) if
+    /// dont_kill is unset and one of these cases is hit:
+    /// 1. the versions do not match
+    /// 2. the server is not running
+    /// 3. the server is unresponsive
+    pub async fn connect(self) -> Result<DaemonClient<DaemonConnector>, DaemonError> {
+        for _ in 0..Self::CONNECT_RETRY_MAX {
+            let pid = self.get_or_start_daemon().await?;
+            debug!("got daemon with pid: {}", pid);
+
+            let path = match self.wait_for_socket().await {
```

Comment:
I _think_ there is a race condition here.

Does the `wait_for_file` call come after `get_or_start_daemon`? Is it possible that we could end up waiting for something to happen separately from this process?

Also, what's the latency and overhead cost on `wait_for_file` vs. just optimistically reading it after it should already be there?

Reply:
The go implementation just used a backoff for this, but I wanted to get notify in the codebase somewhere ahead of the file watcher stuff.

inotify / kqueue and friends should be pretty fast. The server takes up to a few 100ms on my machine to set up the socket so I would assume that there is more IO activity from polling than setting up a watcher. I think when the daemon server lands I will refactor this a little more into a common utility.

Reply:
Notify also supports a polling instead which should be pretty much identical to what we had on the go side, but with an extra runtime cost.